helios-services/src/main/java/com/spotify/helios/master/ZooKeeperMasterModel.java

```diff
@@ -433,11 +433,72 @@ public void removeDeploymentGroup(final String name) throws DeploymentGroupDoesN
     }
   }
 
+  /**
+   * Determines whether we should allow deployment group hosts to be updated.
+   *
+   * We don't want deployment groups that are ROLLING_OUT to change hosts in order to avoid the
+   * following race:
+   *
+   * - Manual rolling update A of a bad, broken job is triggered.
+   * - The hosts change before rolling update A completes, triggering rolling update B.
+   * - Rolling update B fails, because it inherited the bad, broken job from update A.
+   * - The hosts change again, and we trigger unwanted rolling update C because the the previous
+   *   rolling update (B) had reason=HOSTS_CHANGED.
+   *
+   * It's safe to simply abort the hosts changed operation as updateOnHostsChange will be called
+   * again by a reactor.
+   *
+   * @param status The status of the deployment group attempting to change hosts.
+   * @return True if it's safe to update the hosts.
+   */
+  private boolean allowHostChange(final DeploymentGroupStatus status) {
+    if (status == null) {
+      // We're in an unknown state. Go hog wild.
+      return true;
+    }
+    return status.getState() != ROLLING_OUT;
+  }
+
+  /**
+   * Determines whether we should trigger a rolling update when deployment group hosts change.
+   *
+   * We want to avoid triggering an automatic rolling update if the most recent rolling update was
+   * triggered manually, and failed.
+   *
+   * @param group The deployment group that is changing hosts.
+   * @param status The status of the aforementioned deployment group.
+   * @return True if we should perform a rolling update.
+   */
+  private boolean updateOnHostChange(final DeploymentGroup group,
+                                     final DeploymentGroupStatus status) {
+    if (status == null) {
+      // We're in an unknown state. Go hog wild.
+      return true;
+    }
+
+    if (group.getRollingUpdateReason() == null) {
+      // The last rolling update didn't fail, so there's no reason to expect this one will.
+      return status.getState() != FAILED;
+    }
+
+    if (group.getRollingUpdateReason() == HOSTS_CHANGED && status.getState() == FAILED) {
+      // The last rolling update failed, but it was triggered by hosts changing. Try again.
+      return true;
+    }
+
+    // The last rolling update didn't fail, so there's no reason to expect this one will.
+    return status.getState() != FAILED;
+  }
+
   @Override
   public void updateDeploymentGroupHosts(final String groupName, final List<String> hosts)
       throws DeploymentGroupDoesNotExistException {
     log.debug("updating deployment-group hosts: name={}", groupName);
     final ZooKeeperClient client = provider.get("updateDeploymentGroupHosts");
+    final DeploymentGroupStatus status = getDeploymentGroupStatus(groupName);
+    if (!allowHostChange(status)) {
+      return;
```

Comment:
Do you think it's really worth explaining here when it's also explained in the `allowHostChange` javadoc? I considered it but it seemed like it might be excessive.